<?xml version="1.0"?>
<robot
    xmlns:xacro="http://www.ros.org/wiki/xacro" name="robotic_arm" >


    <joint name="joint1" type="revolute">
      <limit effort="100.0" lower="-3.14" upper="3.14" velocity="0.5"/>
      <parent link="main_body"/>
      <child link="link1"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <axis xyz="0 0 1"/>
    </joint>

    <link name="link1">
      <!-- visual & collisiont -->
      <inertial>
        <origin xyz="0 0 -0.05" rpy="0 0 0"/>
        <mass value="0.05"/>
        <inertia
            ixx="4.167e-6" ixy="0.0" ixz="0.0"
            iyy="4.167e-6" iyz="0.0"
            izz="4.167e-7"/>
      </inertial>
      <visual>
        <origin rpy="0.0 0 0" xyz="0.0 0 -0.05"/>
        <geometry>
          <box size="0.01 0.01 0.1"/>
        </geometry>
      </visual>
    </link>

    <joint name="joint2" type="revolute">
      <limit effort="100.0" lower="-1.57" upper="1.57" velocity="0.5"/>
      <parent link="link1"/>
      <child link="link2"/>
      <origin rpy="0 1.57 0" xyz="0 0 -0.1"/>
      <axis xyz="0 1 0"/>
    </joint>

    <link name="link2">
      <!-- visual & collisiont -->
      <inertial>
        <origin xyz="0.05 0 0" rpy="0 0 0"/>
        <mass value="0.05"/>
        <inertia
            ixx="4.167e-7" ixy="0.0" ixz="0.0"
            iyy="4.167e-6" iyz="0.0"
            izz="4.167e-6"/>
      </inertial>
      <visual>
        <origin rpy="0.0 0 0" xyz="0.05 0 0"/>
        <geometry>
          <box size="0.1 0.01 0.01"/>
        </geometry>
      </visual>
    </link>

    <joint name="joint3" type="revolute">
      <limit effort="100.0" lower="-1.57" upper="1.57" velocity="0.5"/>
      <parent link="link2"/>
      <child link="link3"/>
      <origin rpy="0 0 0" xyz="0.1 0 0"/>
      <axis xyz="0 1 0"/>
    </joint>

    <link name="link3">
      <!-- visual & collisiont -->
      <inertial>
        <origin xyz="0.05 0 0" rpy="0 0 0"/>
        <mass value="0.05"/>
        <inertia
            ixx="4.167e-7" ixy="0.0" ixz="0.0"
            iyy="4.167e-6" iyz="0.0"
            izz="4.167e-6"/>
      </inertial>
      <visual>
        <origin rpy="0.0 0 0" xyz="0.05 0 0"/>
        <geometry>
          <box size="0.1 0.01 0.01"/>
        </geometry>
      </visual>
    </link>

    <joint name="joint_eff" type="fixed">
      <parent link="link3"/>
      <child link="end_effector"/>
      <origin rpy="0 0 0" xyz="0.1 0 0"/>
    </joint>


    <link name="end_effector">
     <inertial>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <mass value="0.001"/>
        <inertia
            ixx="0.00" ixy="0.0" ixz="0.0"
            iyy="0.00" iyz="0.0"
            izz="0.00"/>
      </inertial>
    </link>
      <!-- visual & collisiont -->


    <transmission name="joint1_tran">
      <type>transmission_interface/SimpleTransmission</type>
      <joint name="joint1">
        <!-- TODO: effort is torque, maybe position is enough -->
        <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      </joint>
      <actuator name="servo1">
        <!-- TODO: effort is torque, maybe position is enough -->
        <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
        <mechanicalReduction>1</mechanicalReduction>
      </actuator>
    </transmission>

    <transmission name="joint2_tran">
      <type>transmission_interface/SimpleTransmission</type>
      <joint name="joint2">
        <!-- TODO: effort is torque, maybe position is enough -->
        <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      </joint>
      <actuator name="servo2">
        <!-- TODO: effort is torque, maybe position is enough -->
        <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
        <mechanicalReduction>1</mechanicalReduction>
      </actuator>
    </transmission>

    <transmission name="joint3_tran">
      <type>transmission_interface/SimpleTransmission</type>
      <joint name="joint3">
        <!-- TODO: effort is torque, maybe position is enough -->
        <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      </joint>
      <actuator name="servo3">
        <!-- TODO: effort is torque, maybe position is enough -->
        <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
        <mechanicalReduction>1</mechanicalReduction>
      </actuator>
    </transmission>


  









 


</robot>
